--- fs/fat.c
+++ fs/fat.c
@@ -253,6 +253,8 @@
     if (bpb.bytes_per_sector != 512) return;
     
     fat_size = (bpb.sectors_per_fat_16 != 0) ? bpb.sectors_per_fat_16 : bpb.ext.fat32.sectors_per_fat_32;
+    if (fat_size == 0) return; // prevent zero sizing
+    
     uint32_t total_sectors = (bpb.total_sectors_16 != 0) ? bpb.total_sectors_16 : bpb.total_sectors_32;
     root_dir_sectors = ((bpb.root_dir_entries * 32) + 511) / 512;
     
